osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -110,6 +148,126 @@ public void handleGetFirmwareVersionResponse(final List<FirmwareVersionDto> firm
+                .flatMap(Collection::stream)
+                .map(e -> new FirmwareVersion(FirmwareModuleType.forDescription(e.getKey().getDescription()),
+                        e.getValue()))
+                .collect(Collectors.toList());
+
+        // remove the history versions
+        firmwareVersionsToCheck.removeAll(firmwareVersionsInHistory);
+
+        return firmwareVersionsToCheck;
+    }
+
+    public void tryToAddFirmwareVersionToHistory(final String organisationIdentification,
+            final String deviceIdentification, final FirmwareVersion firmwareVersion) {
+
+        final FirmwareModule module = createFirmwareModule(firmwareVersion);
+        final Device device = this.deviceRepository.findByDeviceIdentification(deviceIdentification);
+        final List<FirmwareFile> firmwareFiles = this
+                .getAvailableFirmwareFilesForDeviceModel(organisationIdentification, device.getDeviceModel());
+
+        // check each file for the module and the version as returned by the
+        // device
+        boolean recordAdded = false;
+
+        for (final FirmwareFile file : firmwareFiles) {
+            final Map<FirmwareModule, String> moduleVersions = file.getModuleVersions();
+            if (moduleVersions.containsKey(module) && moduleVersions.get(module).equals(firmwareVersion.getVersion())) {
+
+                // file found, insert a record into the history
+                final DeviceFirmwareFile deviceFirmwareFile = new DeviceFirmwareFile(device, file, new Date(),
+                        INSTALLER);
+                // this.saveDeviceFirmwareFile(deviceFirmwareFile);
+                this.deviceFirmwareFileRepository.save(deviceFirmwareFile);
+                LOGGER.info("Firmware version {} added to device {}", firmwareVersion.getVersion(),
+                        deviceIdentification);
+
+                // we only want to add one record in history
+                recordAdded = true;
+                break;
+            }
+        }
+
+        if (!recordAdded) {
+            LOGGER.warn("No firmware file record found for: {} for device: {}", firmwareVersion, deviceIdentification);
+        }
+
+    }
+
+    private static FirmwareModule createFirmwareModule(final FirmwareVersion firmwareVersion) {
+        final String description = firmwareVersion.getFirmwareModuleType().getDescription().toLowerCase(
+                Locale.getDefault());
+        return new FirmwareModule(description);
+    }
+
+    private List<FirmwareFile> getAvailableFirmwareFilesForDeviceModel(final String organisationIdentification,
+            final DeviceModel deviceModel) {
+        final Manufacturer manufacturer = deviceModel.getManufacturer();
+
+        return this.findAllFirmwareFiles(organisationIdentification, manufacturer.getCode(),
+                deviceModel.getModelCode());
+    }
+
+    private List<FirmwareFile> findAllFirmwareFiles(final String organisationIdentification, final String manufacturer,
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused method parameter "organisationIdentification". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1172)